fdbserver/KeyValueStoreSQLite.actor.cpp

```diff
@@ -92,30 +93,43 @@ struct PageChecksumCodec {
 		int dataLen = pageLen - sizeof(SumType);
 		SumType sum;
 		SumType *pSumInPage = (SumType *)(pData + dataLen);
+		bool usingCrc = false;
+
+		if (write) {
+			// Always write a CRC32 checksum for new pages
+			sum.part1 = 0; // Indicates CRC32 is being used
+			sum.part2 = crc32c_append(0xfdbeefdb, static_cast<uint8_t*>(data), dataLen);
+			*pSumInPage = sum;
+			return true;
+		}
 
-		// Write sum directly to page or to sum variable based on mode
-		SumType *sumOut = write ? pSumInPage : &sum;
-		sumOut->part1 = pageNumber; //DO NOT CHANGE
-		sumOut->part2 = 0x5ca1ab1e;
-		hashlittle2(pData, dataLen, &sumOut->part1, &sumOut->part2);
-
-		// Verify if not in write mode
-		if(!write && sum != *pSumInPage) {
-			if(!silent)
-				TraceEvent (SevError, "SQLitePageChecksumFailure")
-					.error(checksum_failed())
-					.detail("CodecPageSize", pageSize)
-					.detail("CodecReserveSize", reserveSize)
-					.detail("Filename", filename)
-					.detail("PageNumber", pageNumber)
-					.detail("PageSize", pageLen)
-					.detail("ChecksumInPage", pSumInPage->toString())
-					.detail("ChecksumCalculated", sum.toString());
-
-			return false;
+		if (pSumInPage->part1 == 0) {
+			// part1 being 0 indicates with high probability that a CRC32 checksum
+			// was used, so check that first. If this checksum fails, there is still
+			// some chance the page was written with hashlittle2, so fall back to checking
+			// hashlittle2
+			usingCrc = true;
+			sum.part1 = 0;
+			sum.part2 = crc32c_append(0xfdbeefdb, static_cast<uint8_t*>(data), dataLen);
+			if (sum == *pSumInPage) return true;
 		}
 
-		return true;
+		SumType hashLittle2Sum;
+		hashLittle2Sum.part1 = pageNumber; // DO NOT CHANGE
+		hashLittle2Sum.part2 = 0x5ca1ab1e;
+		hashlittle2(pData, dataLen, &hashLittle2Sum.part1, &hashLittle2Sum.part2);
+		if (hashLittle2Sum == *pSumInPage) return true;
+
+		TraceEvent(SevError, "SQLitePageChecksumFailure")
```

Comment:
The trace event only prints one calculated checksum, but if there is a failure then potentially 2 sums were calculated, first CRC32 if part1 is 0 and then hashlittle2 unconditionally.  

Since you have two separate SumType variables, you could log both by splitting ChecksumCalculated into two different details, ChecksumCalculatedHL2 and ChecksumCalculatedCRC32, with the latter only being added if usingCRC is true.